src/main/java/data_gateway/notification/NotificationEntityManager.java

```diff
@@ -0,0 +1,121 @@
+package data_gateway.notification;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import entity.Notification;
+import services.Snowflake;
+import services.notification.NotificationCreationModel;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.FileNotFoundException;
+import java.lang.reflect.Type;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+
+public class NotificationEntityManager implements NotificationManager {
+    private final List<Notification> notifications;
+    private final Gson gson;
+
+    public NotificationEntityManager(Snowflake snowflake) {
+        this.notifications = new ArrayList<>();
+        GsonBuilder builder = new GsonBuilder();
+        builder.registerTypeAdapter(Notification.class, new JsonNotificationAdapter());
+        gson = builder.create();
+    }
+
+    /**
+     * Adds a notification to the database.
+     * @param model the DTO that contains all the information to create a notification
+     */
+    @Override
+    public void addNotification(NotificationCreationModel model) {
+        Notification notification = new Notification(model.getAssociatedId(), model.getTimeInAdvance(),
+                model.getNotificationDateTime(), model.getMessage());
+
+        notifications.add(notification);
+    }
+
+    /**
+     * Deletes a notification
+     * @param associatedId the id of the associated object that the notification is for
+     * @param timeInAdvance the duration in advance that the notification was to be sent out
+     */
+    @Override
+    public void deleteNotification(long associatedId, Duration timeInAdvance) {
+        for (Notification notification : notifications) {
+            if (notification.getAssociatedId() == associatedId
+                    && notification.getNotificationTimeInAdvance().equals(timeInAdvance)) {
+                notifications.remove(notification);
+                break;
+            }
+        }
+    }
+
+    /**
+     * @return all notifications
+     */
+    @Override
+    public List<NotificationReader> getAllNotifications() {
+        List<NotificationReader> notificationReaderList = new ArrayList<>();
+        for (Notification notification : notifications) {
+            notificationReaderList.add(new NotificationToNotificationReader(notification));
+        }
+        return notificationReaderList;
+    }
+
+    /**
+     * Return a notification.
+     * @param associatedId the id of the associated object that the notification is for
+     * @param timeInAdvance the duration in advance that the notification was to be sent out
+     * @return the corresponding notification
+     */
+    @Override
+    public NotificationReader getNotification(long associatedId, Duration timeInAdvance) {
```

Comment:
Why is a notification identified by both its id and time in advance? Is there a reason it can't just be identified by id?

Same question for delete

I guess a more specific question is, you're already injecting in the Snowflake, so why not generate ids for the notifs instead of using the associated id as its own id?

Reply:
^ after some external discussion we decided that notifications would only be accessed in the context of its associated object, hence they should not need their own ids for identification. An associated object may have multiple times, hence the two identifying pieces of data.